From 58f4d33121bc2c60d2ae19ef117fee4cdbbea683 Mon Sep 17 00:00:00 2001
From: Nicolas Le Bayon <nicolas.le.bayon@foss.st.com>
Date: Fri, 4 Apr 2025 17:17:46 +0200
Subject: [PATCH 864/954] fix(st-ddr): skip DDR power init on standby exit for
 STM32MP1

In stm32mp1_ddr_init(), no need to initialize the DDR supplies for
self-refresh exit (standby) as they are not stopped.
Do it only on cold boot sequences.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 39818490c6522962efa18cacf291d5714b3bbd11 fix(st-ddr): skip DDR power init on standby exit for STM32MP1]
Signed-off-by: Nicolas Le Bayon <nicolas.le.bayon@foss.st.com>
Change-Id: Ie2713c4e72c48b131f7aad3fba4c16453a2d8f85
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/449487
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
Tested-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
---
 drivers/st/ddr/stm32mp1_ddr.c | 28 +++++++++++++++-------------
 1 file changed, 15 insertions(+), 13 deletions(-)

diff --git a/drivers/st/ddr/stm32mp1_ddr.c b/drivers/st/ddr/stm32mp1_ddr.c
index 16120ce6f6..51d47507cf 100644
--- a/drivers/st/ddr/stm32mp1_ddr.c
+++ b/drivers/st/ddr/stm32mp1_ddr.c
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2018-2024, STMicroelectronics - All Rights Reserved
+ * Copyright (C) 2018-2025, STMicroelectronics - All Rights Reserved
  *
  * SPDX-License-Identifier: GPL-2.0+ OR BSD-3-Clause
  */
@@ -662,19 +662,21 @@ void stm32mp1_ddr_init(struct stm32mp_ddr_priv *priv,
 	uint64_t time;
 	int ret = -EINVAL;
 
-	if ((config->c_reg.mstr & DDRCTRL_MSTR_DDR3) != 0U) {
-		ret = stm32mp_board_ddr_power_init(STM32MP_DDR3);
-	} else if ((config->c_reg.mstr & DDRCTRL_MSTR_LPDDR2) != 0U) {
-		ret = stm32mp_board_ddr_power_init(STM32MP_LPDDR2);
-	} else if ((config->c_reg.mstr & DDRCTRL_MSTR_LPDDR3) != 0U) {
-		ret = stm32mp_board_ddr_power_init(STM32MP_LPDDR3);
-	} else {
-		ERROR("DDR type not supported\n");
-	}
+	if (!config->self_refresh) {
+		if ((config->c_reg.mstr & DDRCTRL_MSTR_DDR3) != 0U) {
+			ret = stm32mp_board_ddr_power_init(STM32MP_DDR3);
+		} else if ((config->c_reg.mstr & DDRCTRL_MSTR_LPDDR2) != 0U) {
+			ret = stm32mp_board_ddr_power_init(STM32MP_LPDDR2);
+		} else if ((config->c_reg.mstr & DDRCTRL_MSTR_LPDDR3) != 0U) {
+			ret = stm32mp_board_ddr_power_init(STM32MP_LPDDR3);
+		} else {
+			ERROR("DDR type not supported\n");
+		}
 
-	if (ret != 0) {
-		ERROR("DDR power init failed\n");
-		panic();
+		if (ret != 0) {
+			ERROR("DDR power init failed\n");
+			panic();
+		}
 	}
 
 	VERBOSE("name = %s\n", config->info.name);
-- 
2.53.0

